### ISSUE TITLE
[RFC] vllm-spyre-next

### ISSUE THREAD

**Issue by tdoublep** [torch-spyre/spyre-inference#130]
# vllm-spyre-next

This document describes the planned changes to vllm-spyre to enable the next-generation of the Spyre software stack. 

## TL;DR

Today, the [vllm-spyre](https://github.com/vllm-project/vllm-spyre.git) plugin re-uses relatively little of upstream vLLM, and introduces a lot of custom code. By taking advantage of planned efforts to integrate Spyre more deeply into PyTorch, we aim to massively reduce the footprint of the Spyre plugin for vLLM. This will allow us to take advantage of vLLM's diverse model support and its complete suite of performance optimizations, while at the same time reducing the amount of custom code that needs to be maintained.

## History and where we are today

Due to limitations of the current Spyre software stack, the vLLM Spyre plugin need to inject custom scheduling logic, a custom model runner, and custom modeling code and paged attention implementation via [FMS](https://github.com/foundation-model-stack/foundation-model-stack). 

This architecture was the **right** choice when we started development in June 2024 for three key reasons:
1. Spyre integrates into PyTorch through torch.compile, but vLLM's modeling code, was not compatible with torch.compile at that time.
2. The Spyre compiler/runtime imposes constraints which meant that it was not feasible to re-use vLLM's memory management, scheduling and batching logic.
3. The relatively small surface area of interaction between vLLM and the plugin protected us from breaking changes coming from upstream.

Over the last 18 months, both (1) and (2) have changed dramatically. 

Regarding (1): vLLM is now a "frontier user" of torch.compile. It is enabled by default and Meta works closely with the vLLM team to develop custom features and integrations specifically for vLLM. Some examples of these features include graph partitioning, caching of compiled artifacts, and custom guard management. 

Regarding (2): the IBM team has recently open-sourced the [torch-spyre](https://github.com/torch-spyre/torch-spyre) project along with an aggressive roadmap for development. This project aims to enable Spyre as a first-class citizens in PyTorch. It will allow vLLM to manage torch tensors that reside directly on the Spyre device, and provide the possibility to run vLLM on Spyre in eager mode, massively easing development. But perhaps most importantly, it integrates with torch.compile at a deeper level. Before torch-spyre, Spyre integrates with PyTorch at the level of Dynamo: we use Dynamo to trace the modeling code, producing an FX graph which is then consumed by our compiler/runtime. With torch-spyre, we aim to provide a native Spyre backend for Inductor, thus re-using significantly more of the torch.compile stack. It is hoped that this will remove many of the constraints on scheduling and batching that necessitated much of the complexity in the Spyre plugin. 

### Now

The architecture of the current design point is illustrated below:

<img width="2160" height="1214" alt="Image" src="https://github.com/user-attachments/assets/8f8a0972-49ee-4d7e-9db4-125c1109dc80" />

We make the following observations:

- We require a custom scheduler (`SpyreScheduler`) and a custom worker (`SpyreWorker`) and model runner (`SpyreModelRunner`).
- The `SpyreWorker` consumes both modeling code and attention implementation from an external dependency (`fms`).
- While torch.compile is used, a custom backend (`sendnn`) is needed. 
- We do not use any of the modeling code from vLLM.
- We do not use any of vLLM's custom torch.compile features and integration points. . 

### Future

We architecture we would like to build is illustrated below:

<img width="2286" height="1264" alt="Image" src="https://github.com/user-attachments/assets/544ef6e1-6dbd-46b0-b873-69c1f8bf6be5" />

We make the following observations:

- We no longer require a custom scheduler.
- We assume it will be necessary to maintain a Spyre-specific worker and model runner (since all other hardware plugins do this to the best of our knowledge). Ideally in the future common behavior could be factored out as part of an upstream refactor.
- There is no more dependency on `fms`. 
- The `SpyreModelRunner` will now consume the vLLM modeling code directly, with a custom attention backend (`SpyreAttentionBackend`).
- Torch compile will now be used with the Inductor backend (targeting Spyre device).
- Where it makes sense, we will re-use vLLM's custom torch.compile features and integration points. 

## Workstreams

We enumerate the key pieces of work with respect to the blue labeling in the figure above. 

### 1. Spyre Attention Backend 

Tracking issue: torch-spyre/spyre-inference#119 

On GPUs today, vLLM traces the graph of the model and *splits* it into two sub-graphs, one sub-graph containing the attention (and mamba) layers, and another containing everything else. The latter sub-graph is then passed to Inductor which will perform optimizations (e.g., operator fusion etc) on the modeling code. The former sub-graph is *not* passed to Inductor at all. The reason for this is that vLLM includes highly optimized CUDA and Triton kernels that implement paged attention and mamba, and these are more efficient than what Inductor would produce. 

On Spyre, we do not plan to support Triton kernels, so we will need to write PyTorch-native attention and mamba backends for vLLM. For attention, this may look something like the implementation that exists inside FMS today. However, significant work will be needed to adapt that code to work with vLLM block management and data structures.

In order to enable Inductor (and the stack below) to produce performant binaries for execution on Spyre, it is expected that we will need to provide additional hints which may take the form of custom annotations to the PyTorch code. There is also a possibility these annotations may take on a more structured form via some kind domain-specific language (e.g., Helion) but this is under exploration and not yet plan of record. 

### 2. Coverage of vLLM modeling code 

Tracking issue: torch-spyre/spyre-inference#125 

We assume that the vast majority of vLLM modeling code will not be supported initially with the new stack. We will initially target a few key models that currently are not supported on the old stack (e.g., `granite-4.0-h-small`) and work to figure out what needs to be enabled within torch-spyre and the lower-levels of the stack to support them. This work with be performed in close collaboration with the squad working on Inductor for Spyre, as well as the squad enabling the underling ops. 

### 3. Compatibility with custom entrypoints

Tracking issue: torch-spyre/spyre-inference#116 

Over the last year, we have learned firsthand the importance of caching of compiled artifacts to ensure fast start-up times for vLLM on Spyre. It was necessary to implement our own custom caching layer within `torch_sendnn` to be able to handle this. As part of this redesign, we assume that we will re-use the logic that upstream vLLM maintains for caching of compiled artifacts. This workstream involves ensuring that vLLM's custom compilation features play nicely with torch-spyre, and may involve contributing code upstream to enable this.

### General comments

It is expected that all of the above workstreams will involve deep expertise in how vLLM works, but also a willingness to dig into how Inductor and torch-spyre work. The vLLM squad are expected to be able to help debug issues in the underlying stack that may arise from any of the above workstream. 

## The user perspective 

From a user perspective, **nothing will change** in terms of how they use vLLM on Spyre. All user-facing APIs remain the same. Users will benefit from broader model support and should expect performance improvements for workloads that experience severe queuing delays with the current stack. 

## Risks

In the interest of transparency, we also document what we feel are the risks associated with this plan:

**Attention/mamba performance**

vLLM does not currently expose anything like a PyTorch-native attention backend for use with GPUs today. The closest thing is probably the FlexAttention backend, but this is using Triton templates under the hood. Today, it is widely accepted that Inductor cannot produce highly optimized attention implementations by itself (e.g., the compiler today cannot discover the online softmax algorithm from FlashAttention-2). Thus, we find it somewhat risky to assume that Inductor (with whatever improvements we enhance it with) will turn PyTorch-native code into performant binaries for execution Spyre.  

**Breaking changes from upstream**

Deeper integration with vLLM will inevitably expose us more to breaking changes in the plugin API coming from upstream. This is a common complaint from the maintainers of other vLLM hardware plugins that already integrate at this level. Our hope is that we can step up and take a leadership role in the community to try to help stabilize the plugin interface to minimize this risk.

**Comment by maxdebayser** [comment]
> We no longer require a custom scheduler.

I think this will depend on whether we can bridge the gap between vllm's default block size and the chunk size that has the best performance on Spyre hardware for chunked prefill. Maybe we could add this risk: "Need for custom scheduler".

**Comment by maxdebayser** [comment]
> Breaking changes from upstream

As a mitigation, our strategy of defining tests just for tracking the compatibility code we add in the plugin has served is well so far.
